File: c26_2_trie_structure.py
import unittest

class Trie:
    def __init__(self):
        self.v=[{}]
        self.last=set()
    def add(self,s):
        cur=0
        for c in s:
            if c in self.v[cur]:
                cur=self.v[cur][c]
            else:
                self.v[cur][c]=len(self.v)
                cur=len(self.v)
                self.v.append({})
        self.last.add(cur)
    def __contains__(self,s):
        cur=0
        for c in s:
            if c not in self.v[cur]:
                return False
            cur=self.v[cur][c]
        return cur in self.last
    def prefix(self,s):
        l,cur=0,0
        for i,c in enumerate(s):
            if c not in self.v[cur]:
                return s[:l]
            cur=self.v[cur][c]
            if cur in self.last:
                l=i+1
        return s[:l]

class Test(unittest.TestCase):
    def test(self):
        t=Trie()
        t.add('CANAL')
        t.add('CANDY')
        t.add('THE')
        t.add('THERE')
        self.assertFalse('THERN' in t)
        self.assertTrue('THE' in t)
        self.assertEqual(t.prefix('THERN'),'THE')

if __name__=='__main__':
    unittest.main()

File: c++/c26_2_trie_structure.cpp
#include <bits/stdc++.h>
#include <gtest/gtest.h>
using namespace std;
using ll = long long;

namespace {
	struct Trie {
		Trie() : v(1), last() {}
		void add(const string& s) {
			ll cur = 0;
			for (char c : s) {
				if (v[cur].contains(c))
					cur = v[cur][c];
				else {
					v[cur][c] = ssize(v);
					cur = ssize(v);
					v.resize(v.size() + 1);
				}
			}
			last.insert(cur);
		}
		bool contains(const string& s) {
			ll cur = 0;
			for (char c : s) {
				if (!v[cur].contains(c))
					return false;
				cur = v[cur][c];
			}
			return last.contains(cur);
		}
		string prefix(const string& s) {
			ll l = 0, cur = 0;
			for (ll i = 0; i < ssize(s); ++i) {
				if (!v[cur].contains(s[i]))
					return s.substr(0, l);
				cur = v[cur][s[i]];
				if (last.contains(cur))
					l = i + 1;
			}
			return s.substr(0, l);
		}
	private:
		vector<unordered_map<char, ll>> v;
		unordered_set<ll> last;
	};
}

TEST(C262TrieStructure, Trie) {
	Trie t;
	t.add("CANAL");
	t.add("CANDY");
	t.add("THE");
	t.add("THERE");
	EXPECT_FALSE(t.contains("THERN"));
	EXPECT_TRUE(t.contains("THE"));
	EXPECT_EQ(t.prefix("THERN"), "THE");
}
